frontend/src/scenes/events/eventDefinitionsLogic.ts

```diff
@@ -0,0 +1,98 @@
+import { eventDefinitionsLogicType } from './eventDefinitionsLogicType'
+
+interface EventDefinitionStorage {
+    count: number
+    next: null | string
+    results: EventDefinition[]
+}
+
+interface EventsGroupedInterface {
+    label: string
+    options: SelectOption[]
+}
+
+export const eventDefinitionsLogic = kea<
+    eventDefinitionsLogicType<EventDefinitionStorage, EventDefinition, EventsGroupedInterface, SelectOption>
+>({
+    reducers: {
+        eventStorage: [
+            { results: [], next: null, count: 0 } as EventDefinitionStorage,
+            {
+                loadEventDefinitionsSuccess: (state, { eventStorage }) => {
+                    return {
+                        count: eventStorage.count,
+                        results: [...state.results, ...eventStorage.results],
+                        next: eventStorage.next,
+                    }
+                },
+            },
+        ],
+    },
+    loaders: ({ values }) => ({
+        eventStorage: [
+            { results: [], next: null, count: 0 } as EventDefinitionStorage,
+            {
+                loadEventDefinitions: async (initial?: boolean) => {
+                    const url = initial ? 'api/projects/@current/event_definitions/' : values.eventStorage.next
+                    if (!url) {
+                        throw new Error('Incorrect call to eventDefinitionsLogic.loadEventDefinitions')
+                    }
+                    return await api.get(url)
+                },
+            },
+        ],
+    }),
+    listeners: ({ actions }) => ({
+        loadEventDefinitionsSuccess: ({ eventStorage }) => {
+            if (eventStorage.next) {
+                actions.loadEventDefinitions()
+            }
+        },
+    }),
+    events: ({ actions }) => ({
+        afterMount: () => {
+            actions.loadEventDefinitions(true)
+        },
+    }),
+    selectors: {
+        loaded: [
+            // Whether *all* the event definitions are fully loaded
+            (s) => [s.eventStorage, s.eventStorageLoading],
+            (eventStorage, eventStorageLoading): boolean => !eventStorageLoading && !eventStorage.next,
+        ],
+        eventDefinitions: [(s) => [s.eventStorage], (eventStorage): EventDefinition[] => eventStorage.results],
+        eventNames: [
+            // TODO: This can be improved for performance by enabling downstream components to use `eventDefinitions` directly and getting rid of this selector.
+            (s) => [s.eventDefinitions],
+            (eventDefinitions): string[] => eventDefinitions.map((definition) => definition.name),
+        ],
+        customEventNames: [
+            (s) => [s.eventNames],
+            (eventNames): string[] => eventNames.filter((event) => !event.startsWith('!')),
+        ],
+        eventNamesGrouped: [
+            // TODO: This can be improved for performance by enabling downstream components to use `eventDefinitions` directly and getting rid of this selector.
```

Comment:
Again, the point of selectors is to actually improve performance. Here things are calculated just once (and only the first time this value is used)... whereas in React unless you do a lot of memoization, things get recalculated on every render, which is slower. So it makes sense to use as many selectors as feasible.

Reply:
I removed the TODO comments as I think it's totally fine to use as many selectors as you want.